asm/kanread.s

```diff
@@ -0,0 +1,209 @@
+
+// assembler directives
+.set noat
+.set noreorder
+
+.section .text
+
+.balign 16
+
+/**
+ * s32 Kanji_OffsetFromShiftJIS(s32 shiftJISCodepoint);
+ *
+ * Returns the offset of the glyph texture data in the file `kanji` corresponding
+ * to a given 2-byte Shift-JIS codepoint. No range validity check is carried out.
+ *
+ * A nice Shift-JIS codepoint table: https://uic.io/en/charset/show/shift_jis/
+ * The file `kanji` contains the 'Level 1' kanji (0x889F-0x9872), and a reworked
+ * version of the non-kanji section that includes extra English and Hylian glyphs.
+ *
+ * @param shiftJISCodepoint Codepoint of glyph.
+ * return s32 offset into `kanji` file.
+ *
+ * @remark Original name: "LeoGetKadr"
+ */
+LEAF(Kanji_OffsetFromShiftJIS)
+    // Characters with codepoints >= 0x8800 are kanji. Arrangement is regular, 
+    // so convert index directly.
+    li      $at, 0x8800
+    slt     $at, $a0, $at
+    bnez    $at, .nonkanji
+     // 0xBC is number of glyphs in one block in the `kanji` file:
+     // 0x100 possible codepoints with the same byte1
+     // - 0x40 unused at beginning 
+     // - 1 unused at 0x7F 
+     // - 3 unused at 0xFD, 0xFE, 0xFF
+     li     $a2, 0xBC
+    // Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x88
+    multu   $a2, $a1
+    // Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    // 0x__7F is always empty and elided in the file, so decrement
+    bnezl   $at, .kanji_lower_halfblock
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.kanji_lower_halfblock:
+    addi    $a3, $a3, 0x30A
+    add     $a3, $a3, $a2
+    jr      $ra
+     sll    $v0, $a3, 7
+     // returns (0x30A + (adjusted byte2) + (adjusted byte1) * 0xBC) * FONT_CHAR_TEX_SIZE
+
+// Non-kanji are arranged with irregular gaps, use the lookup table.
+.nonkanji:
+    // Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x81
+    multu   $a2, $a1
+    // Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xFF
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    // 0x__7F is always empty and elided in the file, so decrement
+    bnezl   $at, .nonkanji_lower_halfblock
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.nonkanji_lower_halfblock:
+    add     $a3, $a3, $a2
+    lui     $a2, %hi(sKanjiNonKanjiIndices)
+    sll     $a3, $a3, 1
+    addiu   $a2, %lo(sKanjiNonKanjiIndices)
+    add     $a3, $a3, $a2
+    lh      $a2, ($a3)
+    jr      $ra
+     sll    $v0, $a2, 7
+     // returns sKanjiNonKanjiIndices[(adjusted byte2) + (adjusted byte1) * 0xBC] * FONT_CHAR_TEX_SIZE
+END(Kanji_OffsetFromShiftJIS)
+
+// Nearly equivalent C code (Equivalent for all valid input, will behave differently on overflow due to the use of `add` in the original):
+// s32 Kanji_OffsetFromShiftJIS(s32 shiftJISCodepoint) {
+//     s32 byte1;
```

Comment:
While we're at it why not write it more naturally, like idk:
```c
extern s16 sKanjiNonKanjiIndices[];

s32 Kanji_OffsetFromShiftJIS(s32 shiftJISCodepoint) {
    s32 byte1 = shiftJISCodepoint >> 8;
    s32 byte2 = shiftJISCodepoint & 0xFF;

    if (shiftJISCodepoint >= 0x8800) {
        byte1 -= 0x88;
        if (byte2 > 0x7F) {
            byte2--;
        }
        byte2 -= 0x40;

        return (0x30A + byte1 * 0xBC + byte2) * FONT_CHAR_TEX_SIZE;
    } else {
        byte1 -= 0x81;
        if (byte2 > 0x7F) {
            byte2--;
        }
        byte2 -= 0x40;

        return sKanjiNonKanjiIndices[byte1 * 0xBC + byte2] * FONT_CHAR_TEX_SIZE;
    }
}
```
(take or leave anything you want from this)

Reply:
I mainly wrote it the way that I did to make it clear how it corresponds to the assembly (since it does a lot of redundant stuff). If I were writing it as "good" C I'd move the byte2 stuff out of the if to avoid repeating it anyway, I think it's better to go one way or the other rather than only alter some of it.

Reply:
if our reasoning for including it is to make it easier to understand, Id say that keeping its resemblance to the assembly is not that important. It would help if someone wants to read the assembly along with the c code to double check your work i guess, but the reasoning seems to mostly be accessibility and ease of use/understanding